### PR TITLE
Add PointCloud spatial distribution

### LINKED ISSUE #3159: Allow sampling source location from a point cloud with discrete probabilities
## Description
A new `SpatialDistribution` that has a list of N possible positions, each with its own probability.  The probabilities are sampled discretely from that PDF to find the locations.

## Alternatives
One alternative is to have N Independent `SpatialPoint` distributions, but it seems less efficient on a number of fronts.

### PR DESCRIPTION
**Pull request by gonuke**
# Description

Allows users to define a list of points in space, each with a different relative intensity, to be sampled discretely. This is a valid `SpatialDistribution` and can be used anywhere that a `SpatialDistribution` is valid.

Fixes #3159 

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: include/openmc/distribution_spatial.h
Comment (gonuke) [inline include/openmc/distribution_spatial.h:146]: Note: this could just as easily be a vector of `Position`s but I wasn't sure how to do that in XML space, so decided to just keep it this way for now.  Happy for any guidance on changing this to a different form.

Could also overload the constructor with a few choices, but not sure of OpenMC design practices.

File: include/openmc/distribution_spatial.h
Comment (gonuke) [inline include/openmc/distribution_spatial.h:159]: I didn't think through normalization.  I don't know/recall if a `DiscreteIndex` is normalized, but I guess that it isn't given that a method like this exists for the `MeshSpatial` class.

File: openmc/stats/multivariate.py
Comment (gonuke) [inline openmc/stats/multivariate.py:802]: Another place to seek OpenMC design best practice about being tolerant to lots of data forms for this data

File: openmc/stats/multivariate.py
Comment (gonuke) [inline openmc/stats/multivariate.py:839]: assuming there is no better way to store a list of points in XML

Comment (gonuke) [reply openmc/stats/multivariate.py:839]: Could imagine flattening this to a single list of  `[[x], [y], [z]]` or a single list of `[ (x,y,z) for zip(x,y,z)]`, but not sure it's a better choice

File: include/openmc/distribution_spatial.h
Comment (gonuke) [inline include/openmc/distribution_spatial.h:162]: could assemble these into a vector of `Position` on initialization, but not sure pros/cons of that choice vs. keeping these separate...

File: openmc/stats/multivariate.py
Comment (pshriwise) [inline openmc/stats/multivariate.py:779]:
```suggestion
    positions: numpy.ndarray
        The points in space to be sampled with shape (N, 3)
```

File: openmc/stats/multivariate.py
Comment (pshriwise) [inline openmc/stats/multivariate.py:879]:
```suggestion
```

File: openmc/stats/multivariate.py
Comment (pshriwise) [inline openmc/stats/multivariate.py:865]: I'm thinking the XML format of the data as a flat list of points (X1, Y1, Z1, X2, Y2, Z2, ...) would make the import export code simpler, and it isn't more or less readable in an XML format (for my brain anyway).

File: include/openmc/distribution_spatial.h
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:162]: One more thought, we should make sure to use a container here so the object owns the memory of these attributes -- `gsl::span` is just a view into memory space and doesn't "own" these values.

File: openmc/stats/multivariate.py
Comment (pshriwise) [inline openmc/stats/multivariate.py:839]:
```suggestion
            subelement.text = ' '.join(str(e) for e in self.positions[..., idx])
```
I think?

**Review by pshriwise — CHANGES_REQUESTED** [review]
Thanks for this snappy PR @gonuke! A useful distribution that we haven't been able to support before!

**Review by paulromano — COMMENTED** [review]
@gonuke Thanks for this PR! Out of curiosity, what is the intended application of this feature?

**Comment by gonuke** [comment]
We have a user who has an existing approximation of a volumetric source by a high density list of isotropic point sources.  They want to use that source for comparison to other simulations.

File: include/openmc/distribution_spatial.h
Comment (gonuke) [reply include/openmc/distribution_spatial.h:146]: It looks like adding an input stream operator for `Position` will allow existing code to work, although not sure about robustness

**Comment by gonuke** [comment]
While it may have been elegant to add an `istream` operator for `Position` the version I added did not round-trip with the existing `ostream` operator and it is more convenient in XML to not use the verbose `ostream` format.

Thus, I've relied on just reading a list of doubles and packing them into `Position` in the reader.

File: openmc/stats/multivariate.py
Comment (pshriwise) [inline openmc/stats/multivariate.py:804]:
```suggestion
            raise ValueError('Unable to interpret source positions')
```

**Review by pshriwise — CHANGES_REQUESTED** [review]
Some small line suggestions here @gonuke. I went ahead and added test that samples the external source distribution on the C++ side and factored out a function in the XML interface for parsing a set of `Position`s from a flat array of values on a node. Hope that's okay!

Thanks again for this addition.

File: include/openmc/distribution_spatial.h
Comment (pshriwise) [reply include/openmc/distribution_spatial.h:159]: It appears it is in the `init_alias` method, but the `integral` method retains the original, potentially un-normalized, sum of the probabilities.

File: openmc/stats/multivariate.py
Comment (pshriwise) [reply openmc/stats/multivariate.py:802]: The trend seems to be storing NumPy arrays for the internal attributes values when in doubt. I'll let @paulromano chime in here though in case his perspective differs.

Comment (gonuke) [reply openmc/stats/multivariate.py:802]: But is that imposed as a requirement? or are we tolerant of other forms?

File: openmc/stats/multivariate.py
Comment (gonuke) [reply openmc/stats/multivariate.py:837]: I'm not sure where/how that's defined?  Do you mean some XML grammar? I'm not sure where to find that...

File: openmc/stats/multivariate.py
Comment (pshriwise) [reply openmc/stats/multivariate.py:800]: Tolerant of other forms I'd say.

File: openmc/stats/multivariate.py
Comment (gonuke) [reply openmc/stats/multivariate.py:835]: Aah... found it, and also found that it is missing docs for the MeshSpatial option, so tried to add that, too.

Comment (pshriwise) [reply openmc/stats/multivariate.py:835]: Excellent. Thank you!!

File: docs/source/io_formats/settings.rst
Comment (pshriwise) [inline docs/source/io_formats/settings.rst:610]: Quick fix

```suggestion
      positions provided in the node, based on the relative strengths provided
```

**Review by pshriwise — COMMENTED** [review]
I just pushed a couple of small changes, but this looks good to me! @paulromano, you mentioned you'd like to have a look at this as well, so I'll give you a little time to review.

File: src/distribution_spatial.cpp
Comment (gonuke) [inline src/distribution_spatial.cpp]: again - this was all modeled on `MeshSpatial` - I'm not sure I agree with this change - there should be no reason to test the size of `strengths` except in the case where it is trying to extract XML data.  Moving this test outside of that block seems like overkill.

File: src/distribution_spatial.cpp
Comment (gonuke) [inline src/distribution_spatial.cpp]: Is this a style question?  FWIW, this was modeled on `MeshSpatial` that initializes this to 1.0 and then only changes if the XML data exists.  If this change happens, we may want to change `MeshSpatial` accordingly.

Comment (pshriwise) [reply src/distribution_spatial.cpp]: I moved this below for size considerations. In the case that there's a "strengths" entry on the XML node we'd be storing two vectors of size `point_cloud_`. I'm happy to make an issue and update `MeshSpatial` as well in a separate PR!

File: src/distribution_spatial.cpp
Comment (pshriwise) [reply src/distribution_spatial.cpp]: The control flow feels better to me in this format - fewer levels of indentation doesn't interrupt the logic of the statements above where `strengths` is populated. Placing such a check at the end of the constructor makes sense to me as regardless of how strengths is created above, this condition should be true. It occurs in the constructor, executed once, and shouldn't be expensive. Happy to reconsider if you feel strongly about the previous control flow.

Comment (pshriwise) [reply src/distribution_spatial.cpp]: Thanks for being on board with the updates pushed here. I have a branch with similar changes for the `MeshSpatial` class, so a similar philosophy is adopted there. I'll create a PR with those (and any other improvements that apply there as well) once this has been merged.

**Comment by gonuke** [comment]
Pinging @paulromano to see if this can move forward?

**Review by paulromano — APPROVED** [review]
Thanks @gonuke! I did a bit of refactoring and simplification. I'm good to merge but before I do, let me know if you object to any of my changes.

**Comment by gonuke** [comment]
All looks good to me - thanks @paulromano

**Review by pshriwise — APPROVED** [review]
Thank you @gonuke!

**Pull request merged**
